src/v/cluster/rm_stm.cc

```diff
@@ -1069,35 +1071,179 @@ ss::future<result<raft::replicate_result>> rm_stm::replicate_seq(
           "first_seq={} of the batch should be less or equal to last_seq={}",
           bid.first_seq,
           bid.last_seq);
-        outcome->set_value();
         co_return errc::generic_tx_error;
     }
 
+    // getting session by client's pid
+    ss::lw_shared_ptr<inflight_requests> session;
+    auto session_it = _inflight_requests.find(bid.pid);
+    if (session_it == _inflight_requests.end()) {
+        session = ss::make_lw_shared<inflight_requests>();
+        _inflight_requests.try_emplace(bid.pid, session);
+    } else {
+        session = session_it->second;
+    }
+
+    // critical section, rm_stm evaluates the request its order and
+    // whether or not it's duplicated
+    // ASSUMING THAT ALL INFLIGHT REPLICATION REQUESTS WILL SUCCEED
+    // if any of the request fail we expect that the leader step down
+    auto u = co_await session->lock.get_units();
+
+    if (!_c->is_leader() || _c->term() != synced_term) {
+        co_return errc::not_leader;
+    }
+
+    if (session->term < synced_term) {
+        // we don't care about old inflight requests because the sync
+        // guarantee (all replicated records of the last term are
+        // applied) makes sure that the passed inflight records got
+        // into _log_state->seq_table
+        session->forget();
+        session->term = synced_term;
+    }
+
+    // checking if the request (identified by seq) is already resolved
+    // checking among the pending requests
+    auto cached_r = session->known_seq(bid.last_seq);
+    if (cached_r) {
+        co_return cached_r.value();
+    }
+    // checking among the responded requests
```

Comment:
log_state's seq info doesn't contain gaps and holds only up to 5 requests. All the inflight requests go into _inflight_requests and it may contains gap (a request which is still being replicated).

Usually seq info gets into _log_state by reading the log and replaying already replicated batches (see apply, apply_data methods). But there may be a delay between a moment a record is replicated and a moment it was read so in closer to the end of replicate_seq we explicitly fill _log_state's seq info by applying data from inflight seq info to speed up this process (when we know for sure that the record was replicated).